public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/QueryAndExpressionsStep.tsx

```diff
@@ -0,0 +1,248 @@
+import { getDefaultOrFirstCompatibleDataSource } from '../../../utils/datasource';
+import { ExpressionEditor } from '../ExpressionEditor';
+import { ExpressionsEditor } from '../ExpressionsEditor';
+import { QueryEditor } from '../QueryEditor';
+import { RuleEditorSection } from '../RuleEditorSection';
+import { refIdExists } from '../util';
+
+import { AlertType } from './AlertType';
+import {
+  addNewDataQuery,
+  addNewExpression,
+  queriesAndExpressionsReducer,
+  removeExpression,
+  rewireExpressions,
+  setDataQueries,
+  updateExpression,
+  updateExpressionRefId,
+  updateExpressionType,
+} from './reducer';
+
+interface Props {
+  editingExistingRule: boolean;
+}
+
+export const QueryAndExpressionsStep: FC<Props> = ({ editingExistingRule }) => {
+  const runner = useRef(new AlertingQueryRunner());
+  const {
+    setValue,
+    getValues,
+    watch,
+    formState: { errors },
+    control,
+  } = useFormContext<RuleFormValues>();
+  const [panelData, setPanelData] = useState<Record<string, PanelData>>({});
+
+  const initialState = {
+    queries: getValues('queries'),
+    panelData: {},
+  };
+  const [{ queries }, dispatch] = useReducer(queriesAndExpressionsReducer, initialState);
+
+  const type = watch('type');
+  const condition = watch('condition');
+  const dataSourceName = watch('dataSourceName');
+
+  const isGrafanaManagedType = type === RuleFormType.grafana;
+  const isCloudAlertRuleType = type === RuleFormType.cloudAlerting;
+  const isRecordingRuleType = type === RuleFormType.cloudRecording;
+
+  const showCloudExpressionEditor = (isRecordingRuleType || isCloudAlertRuleType) && dataSourceName;
+
+  const cancelQueries = useCallback(() => {
+    runner.current.cancel();
+  }, []);
+
+  const runQueries = useCallback(() => {
+    runner.current.run(queries);
+  }, [queries]);
+
+  // whenever we update the queries we have to update the form too
+  useEffect(() => {
+    setValue('queries', queries, { shouldValidate: false });
+  }, [queries, runQueries, setValue]);
+
+  // set up the AlertQueryRunner
+  useEffect(() => {
+    const currentRunner = runner.current;
+
+    runner.current.get().subscribe((data) => {
+      setPanelData(data);
+    });
+
+    return () => currentRunner.destroy();
+  }, []);
+
+  const noCompatibleDataSources = getDefaultOrFirstCompatibleDataSource() === undefined;
+
+  const isDataLoading = useMemo(() => {
+    return Object.values(panelData).some((d) => d.state === LoadingState.Loading);
+  }, [panelData]);
+
+  // data queries only
+  const dataQueries = useMemo(() => {
+    return queries.filter((query) => !isExpressionQuery(query.model));
+  }, [queries]);
+
+  const emptyQueries = queries.length === 0;
+
+  const onUpdateRefId = useCallback(
+    (oldRefId: string, newRefId: string) => {
+      const newRefIdExists = refIdExists(queries, newRefId);
+      // TODO we should set an error and explain what went wrong instead of just refusing to update
```

Comment:
💬 I think this is pretty much the only `TODO` I haven't figured out yet, any ideas / thoughts?

Reply:
For queries, we display a validation message saying `Query name already exists` but for expressions, something weird happens:
![Kapture 2022-09-29 at 09 58 57](https://user-images.githubusercontent.com/6293563/192975492-36561df7-574e-4f10-9f18-39eac979f105.gif)

Reply:
Ah right, this happens because I've forgotten to add `type=button` to the `remove expression` button which makes it think it should be activated when pressing "enter" – really wish this was the default type for buttons 😢